MaintenanceJobDesignDoc.md

```diff
@@ -0,0 +1,140 @@
+objects that exist in a pack-file. This will use the multi-pack-index to test
+containment, so is only bound by the file-deletion operations.
+
+### Pack loose objects that do not appear in a pack-file
+
+We can use `git pack-objects -q <object-dir>/pack/loose-<timestamp>` to create a
+pack-file filled with a list of loose objects. We can pass the list of loose
+objects to this command through stdin. We can choose to batch this operation by
+limiting the number of objects we pack at a time, if we want. We may also want
+to skip this step if the number of loose objects is low (&lt; 1,000? &lt;
+10,000?)
+
+By listing a timestamp in the pack-file name, we can store the "most-recent" 
+time this job was run.
+
+*Note:* While we are thinking about packing loose objects, we can consider
+_deleting_ some loose blobs. It is easier to delete loose blobs than packed
+blobs. If we want to delete "stale" blobs, then we should add a step to this
+job.
+
+Prefetch Pack Job
+---------------
+
+We store a list of prefetch packs as `<object-dir>/pack/prefetch-<timestamp>-<sha>.[pack|idx]`.
+These are downloaded at a regular cadence by the post-fetch job, and could have
+a new pack-file for every hour. In addition to these prefetch packs, some daily
+prefetch packs will appear and occasionally a very large prefetch pack covering
+everything more than 30 days old. This large prefetch pack appears on initial
+clone and sometimes again if a user does not prefetch for 30 days (this occurs 
+less often now that background prefetch exists).
+
+### Delete pack-files with no objects referenced by the multi-pack-index
+
+The multi-pack-index stores a single copy of every packed object. When an object
+appears in multiple pack-files covered by the multi-pack-index, we select the
+pack-file with the most-recent _modified time_ as reported by the filesystem.
+For example, if someone downloads a second large prefetch pack when prefetching
+for the first time in 31 days, then that large prefetch pack is referenced and
+all older prefetch packs are "stale" and not referenced by the multi-pack-index.
+
+We can extend the `git multi-pack-index` builtin to delete pack-files that have
+no objects referenced by the multi-pack-index. This should happen at the same
+time as we update the multi-pack-index since we don't want to have this list of
+pack-files in the multi-pack-index if they don't exist. We also don't want to
+remove them from the multi-pack-index while the pack-files still exist, because
+then a Git process will start referencing those pack-files in the `packed_git`
+linked list.
+
+### Repack the prefetch pack-files
+
+We can reduce the number of prefetch packs and improve file locality by
+combining prefetch packs into new "repacked" pack-files. Use `git show-index
+<object-dir>/pack/prefetch-<timestamp[i]>-<hash[i]>.idx` to list the objects
+in the prefetch packs and pipe that output to `git pack-objects -q
+<object-dir>/pack/repacked-<timestamp>`.

```

Comment:
Any idea how long this would run or how much memory it would use?

Reply:
This is something to investigate during implementation, but I think if we plan to batch the repack (only repack prefetch packs that are smaller than 1GB until the batch is larger than 1GB) then the computation time would be under a minute. Definitely something to try.

I do want to make an extra-careful statement that we are _not_ repacking the whole pack directory into a single pack. We are just taking the many small packs into slightly fewer, slightly bigger packs.